### PR TITLE
PySpark: implementing subtractByKey(), subtract() and keyBy()

### PR DESCRIPTION
**Pull request by AndreSchumacher**

### CODE REVIEW

**Comment by AndreSchumacher** [comment]
Jenkins, test this please.

File: python/pyspark/rdd.py [hunk 1]
```diff
@@ -754,6 +754,43 @@ def cogroup(self, other, numPartitions=None):
...
+        Return each (key, value) pair in C{self} that has no pair with matching key
+        in C{other}.
+
+        >>> x = sc.parallelize([("a", 1), ("b", 4), ("b", 5), ("a", 2)])
+        >>> y = sc.parallelize([("a", 3), ("c", None)])
+        >>> sorted(x.subtractByKey(y).collect())
+        [('b', 4), ('b', 5)]
+        """ 
+        filter_func = lambda tpl: len(tpl[1][0]) > 0 and len(tpl[1][1]) == 0
+        map_func = lambda tpl: [(tpl[0], val) for val in tpl[1][0]]
```
Comment (mateiz) [inline python/pyspark/rdd.py:768]: FYI, when a lambda takes a tuple, you can actually "pattern-match" on the tuple like this:

```
flter_func = lambda (key, vals): len(vals[0]) > 0 and len(vals[1]) == 0
map_func = lambda (key, vals): [(a, val) for val in vals[0])
```

**Comment by mateiz** [comment]
This looks great.. I made a small comment on a way you can make this kind of code cleaner but I've also merged it since it will be nice to have it in 0.8.

File: python/pyspark/rdd.py (hunk 1 above)
Comment (AndreSchumacher) [inline python/pyspark/rdd.py:768]: Thanks. Good point, that's a lot more readable.

**Pull request merged**